### PR TITLE
fix: Wrap semgrep runs in preemptive threads

### PR DESCRIPTION
**Pull request by ajbt200128**
This PR wraps calls to semgrep-core in preemptive threads so that way scanning a long file doesn't hog the scheduler.

### CODE REVIEW

File: src/osemgrep/language_server/requests/Initialize_request.ml [hunk 1]
```diff
@@ -67,6 +67,8 @@ let initialize_server server
           "intellij"
     | _ -> false
   in
+  Lwt_preemptive.init 1 user_settings.jobs (fun msg ->
```
Comment (aryx) [inline src/osemgrep/language_server/requests/Initialize_request.ml:74]: Yep, using Thread.create internally (looking at ~/.opam/.../lwt/unix/lwt_preemptive.ml)

**Review by aryx — APPROVED** [review]
Very nice! Simple!
Now how will this work for LSP.js??

File: src/osemgrep/language_server/requests/Initialize_request.ml (hunk 1 above)
Comment (aryx) [inline src/osemgrep/language_server/requests/Initialize_request.ml:74]: Maybe a small comment here like (* we're using lwt cooperative threads in most of the code of the language server,
but we still need sometimes to execute code using Unix. or Common. that access files in a blocking way. However
we can wrap those calls in the Lwt_preemptive library which internally use Thread to at least not block
the whole lwt event loop but just block threads in a pool. This init calls creates such a thread pool internally.

File: src/osemgrep/language_server/scan_helpers/Scan_helpers.ml [hunk 2]
```diff
@@ -64,56 +60,66 @@ let run_semgrep ?(targets = None) ?(rules = None) ?(git_ref = None)
...
+      let files = Session.scanned_files server.session in
+      Diagnostics.diagnostics_of_results ~is_intellij:server.session.is_intellij
+        results files
+    in
+    end_progress server token;
+    batch_notify server diagnostics
   in
-  end_progress server token;
-  batch_notify server diagnostics
+  wrap_with_detach f
```
Comment (aryx) [inline src/osemgrep/language_server/scan_helpers/Scan_helpers.ml:81]: maybe a small comment here explaining why we need the detach, what are the Unix/Common blocking calls in f()
(I guess the call to 'run_semgrep' which internally use a lot Unix)

File: src/osemgrep/language_server/scan_helpers/Scan_helpers.ml [hunk 3]
```diff
@@ -64,56 +60,66 @@ let run_semgrep ?(targets = None) ?(rules = None) ?(git_ref = None)
...
+    let files = [ file ] in
+    Session.record_results server.session results files;
+    let diagnostics =
+      Diagnostics.diagnostics_of_results ~is_intellij:server.session.is_intellij
+        results files
+    in
+    batch_notify server diagnostics
   in
-  batch_notify server diagnostics
+  wrap_with_detach f
```
Comment (aryx) [inline src/osemgrep/language_server/scan_helpers/Scan_helpers.ml:126]: same here, small comment explaining what could take a long time in f() and block the whole lwt event loop.

**Review by aryx — APPROVED** [review]

**Review by tpetr — APPROVED** [review]

File: src/osemgrep/language_server/requests/Initialize_request.ml (hunk 1 above)
Comment (ajbt200128) [reply src/osemgrep/language_server/requests/Initialize_request.ml:74]: Makes sense

Comment (ajbt200128) [reply src/osemgrep/language_server/requests/Initialize_request.ml:74]: Will do

**Pull request merged**